### PR TITLE
Handle FILTHY tag when unloading/reloading items

### LINKED ISSUE #29614: Wooden needle (and probably others) transform filthy thread into clean thread
**Describe the bug**  
Loading a wooden needle with filthy thread then unloading it causes the thread to come out clean.

**To Reproduce**  
Get filthy thread.
load wooden needle with it
unload needle

**Expected behavior**  
IDK you tell me :)

**Versions and configuration(please complete the following information):**
Latest version using the launcher 8786

### PR DESCRIPTION
**Pull request by ZhilkinSerg**
#### Summary

```SUMMARY: Bugfixes "Handle FILTHY tag when unloading/reloading items"```

#### Purpose of change

Fixes #29614

#### Describe the solution

Handle `FILTHY` tag when unloading/reloading items:

- make ammo filthy when unloading from filthy item;
- make item filthy when reloading with filthy ammo.

#### Additional context

Maybe we should add some confirmations when trying to reload an item with filthy ammo?

### CODE REVIEW

File: src/activity_handlers.cpp [hunk 1]
```diff
@@ -1879,13 +1879,20 @@ void activity_handlers::reload_finish( player_activity *act, player *p )
...
     if( !reloadable.reload( *p, std::move( act->targets[ 1 ] ), qty ) ) {
         add_msg( m_info, _( "Can't reload the %s." ), reloadable.tname() );
         return;
     }
 
     std::string msg = _( "You reload the %s." );
+
+    if( ammo_is_filthy && !reloadable_is_filthy ) {
+        reloadable.set_flag( "FILTHY" );
+    }
```
Comment (Night-Pryanik) [inline src/activity_handlers.cpp]: Check for `!reloadable_is_filthy` is redundant. If it's already filthy, nothing changes.

**Comment by Funguss** [comment]
Can all items be cleaned now? I remember disassembling some filthy stuff a couple of months back and not being able to clean some of it in either the dishwasher or the washing machine. Haven't had the opportunity to check it out since.

File: src/activity_handlers.cpp [hunk 2]
```diff
@@ -1879,13 +1879,20 @@ void activity_handlers::reload_finish( player_activity *act, player *p )
...
+    const bool ammo_is_filthy = ammo.is_filthy();
 
     if( !reloadable.reload( *p, std::move( act->targets[ 1 ] ), qty ) ) {
         add_msg( m_info, _( "Can't reload the %s." ), reloadable.tname() );
         return;
     }
 
     std::string msg = _( "You reload the %s." );
+
+    if( ammo_is_filthy && !reloadable_is_filthy ) {
```
Comment (kevingranade) [inline src/activity_handlers.cpp]:
```suggestion
    if( ammo_is_filthy ) {
```

**Comment by Amariithynar** [comment]
...This is silly. A magazine loaded into a gun is at most going to be dirty on the bottom, which is easily wipeable. The bullets in the magazine are not going to be encrusted with the filth coating the body of the gun. Similarly, if a flask or bottle is filthy on the outside, the stuff inside isn't going to suddenly become full of all the gunk encrusting it.

**Comment by ZhilkinSerg** [comment]
Yeah, filthy items are easily washable. What's your point?

**Comment by Funguss** [comment]
On the topic of filthy liquids from containers, I initially felt the same, but would you drink from a filthy container? Would you trust something poured from a filthy container? Both situations would end up with the liquid contaminated. I don't believe FILTHY has any kind of effect on consumables at the moment anyway, but I could be wrong.

Hard items can be cleaned with a rag or a sponge, both of which are easily found in a house. Perhaps some exposition when unloading or reloading hard items would be a good idea? Or would that just break immersion?

**Comment by Amariithynar** [comment]
> Yeah, filthy items are easily washable. What's your point?

With clothes it makes sense. Not with bottles that have corks or stoppers or twist caps or the like, and definitely not with magazines, as I already mentioned. The contents won't suddenly inherit all the filth that is on the entirety of the outside of the container, plain and simple.


> On the topic of filthy liquids from containers, I initially felt the same, but would you drink from a filthy container? Would you trust something poured from a filthy container? Both situations would end up with the liquid contaminated. I don't believe FILTHY has any kind of effect on consumables at the moment anyway, but I could be wrong.
> 
> Hard items can be cleaned with a rag or a sponge, both of which are easily found in a house. Perhaps some exposition when unloading or reloading hard items would be a good idea? Or would that just break immersion?

I've done just that while camping, where my thermos fell into some sticky, slimy, definitely-full-of-algae-and-bacteria mud that stank to high heaven. Wiping it off, it was still caked in slimy crud, but I could still pop off the cover and easily unscrew the cap and drink from the completely clean, threaded mouth of the thermos. Sure, I rinsed the outside down afterwards when we passed by a river, but I didn't feel any risk in drinking from it, because of how such containers are designed. Same when I've dropped bottles or jars (either because of my own clumsiness, or they've slid out of pockets or holes in shopping bags) into mud or muddy puddles (and on more than one occasion oil-slicked puddles of murky, dirty water) throughout my life; the cap protects the threaded mouth of the bottle/jar, and keeps it clean so no matter how dirty the outside, you can still drink/eat from it without a worry (especially if the cap wasn't unsealed before for plastic bottles, as the little plastic ring around the mouth that the cap attaches to acts as an even greater splash/dirt guard). Drink cans, though, I definitely don't open and drink from unless the can lid/mouth was completely clear of what it fell into, I'll give you that. Food cans, I'd wash before consuming if they were encrusted in filth, definitely, for the same reasons.

Regardless of anything else, if this is going to stay, there definitely should be a warning about getting 'ammo' filthy that wasn't filthy to begin with, when unloading it from a filthy container.

**Comment by ZhilkinSerg** [comment]
Who says ammo was not filthy initially?

What will this warning achieve?

**Comment by Funguss** [comment]
Good points about the specific vessels. I was only really thinking about one time a bar I knew of had a sewage leak from the toilets above their storeroom. They wiped the bottles down (not very thoroughly) and sold them anyway. Maybe a flag for containers that would keep food safe for consumption?

Regarding ammo/firearms I know precious little, so I can't really comment.

**Comment by Amariithynar** [comment]
Mhm. Any container that has a cap or cover for the mouth would be safe to decant into and from without washing.

https://www.youtube.com/watch?v=NqCLTC_q4h8 skip ahead through the first 35 seconds or so which are more of an ad for the channel, to get to the topic at hand. You see the magazine loaded in the rifle, and then by 0:37 you see he's removed it, and can see that almost half the magazine is tightly seated inside the rifle with a plastic lip that seats against the bottom of the mag well to seal it against dirt, filth, and other debris. There is no access to the ammo inside except through the top port by which the ammo is both loaded into the mag and also ejected into the rifle chamber in preparation to be fired. Even if the rifle was dropped into rotting viscera and completely coated over with it, you could eject the mag and recover the clean ammunition inside without issue. The mag being filthy wouldn't give that status to any other rifle or to any ammo loaded into it, before or afterwards, simply because of the way they're constructed. Pistol magazines are much the same, though a lot more of the magazine seats into the magazine well; for properly sized mags, usually only the bottom plate (which acts as a butt plate for the grip in general as well as to help seal the mag well while the magazine is inside) is visible, though for oversized mags (drum mags, extended mags, etc) the excess length and/or girth is also outside the well, obviously.

**Pull request merged**